core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@junrao Yeah you are right, this doesn't explain the saving. Here is another explanation based on the observation that broker reads requests from all active channels in a round-robin manner. Say there are 10 brokers/clients which are actively sending requests to this broker. Also assume that there is 1 network thread and 1 io thread. At the time broker finishes processing that 3000 requests from controller, this broker would have processed 3000*10 requests in total. With this patch, controller only needs to send 1 LeaderAndIsrRequest, 1 MetadataUpdateRequest and 1 StopReplicaRequest before that ControlledShutdownResponse. On average broker only needs to process a total of 3*10 requests before it touches ControlledShutdownResponse.